Source/Core/DiscIO/DirectoryBlob.cpp

```diff
@@ -0,0 +1,496 @@
+{
+  std::vector<File::FSTEntry> sorted_entries = parent_entry.children;
+
+  // Sort for determinism
+  std::sort(sorted_entries.begin(), sorted_entries.end(), [](const File::FSTEntry& one,
+                                                             const File::FSTEntry& two) {
+    const std::string one_upper = ASCIIToUppercase(one.virtualName);
+    const std::string two_upper = ASCIIToUppercase(two.virtualName);
+    return one_upper == two_upper ? one.virtualName < two.virtualName : one_upper < two_upper;
+  });
+
+  for (const File::FSTEntry& entry : sorted_entries)
+  {
+    if (entry.isDirectory)
+    {
+      u32 entry_index = *fst_offset / ENTRY_SIZE;
+      WriteEntryData(fst_offset, DIRECTORY_ENTRY, *name_offset, parent_entry_index,
+                     entry_index + entry.size + 1, 0);
+      WriteEntryName(name_offset, entry.virtualName);
+
+      WriteDirectory(entry, fst_offset, name_offset, data_offset, entry_index);
+    }
+    else
+    {
+      // put entry in FST
+      WriteEntryData(fst_offset, FILE_ENTRY, *name_offset, *data_offset, entry.size,
+                     m_address_shift);
+      WriteEntryName(name_offset, entry.virtualName);
+
+      // write entry to virtual disc
+      auto result = m_virtual_disc.emplace(*data_offset, entry.size, entry.physicalName);
+      _dbg_assert_(DISCIO, result.second);  // Check that this offset wasn't already occupied
+
+      // 32 KiB aligned - many games are fine with less alignment, but not all
+      *data_offset = Common::AlignUp(*data_offset + std::max<u64>(entry.size, 1ull), 0x8000ull);
+    }
+  }
+}
+
+const DirectoryBlobReader::DiscContent&
+DirectoryBlobReader::AddFileToContents(std::set<DiscContent>* contents, const std::string& path,
+                                       u64 offset, u64 max_size)
+{
+  return *(contents->emplace(offset, std::min(File::GetSize(path), max_size), path).first);
+}
+
+static u32 ComputeNameSize(const File::FSTEntry& parent_entry)
+{
+  u32 name_size = 0;
+  for (const File::FSTEntry& entry : parent_entry.children)
+  {
+    if (entry.isDirectory)
+      name_size += ComputeNameSize(entry);
+
+    name_size += (u32)entry.virtualName.length() + 1;
+  }
+  return name_size;
+}
+
+static void ConvertUTF8NamesToSHIFTJIS(File::FSTEntry& parent_entry)
+{
+  for (File::FSTEntry& entry : parent_entry.children)
+  {
+    if (entry.isDirectory)
+      ConvertUTF8NamesToSHIFTJIS(entry);
+
+    entry.virtualName = UTF8ToSHIFTJIS(entry.virtualName);
+  }
+}
+
+static std::string ASCIIToUppercase(std::string str)
+{
+  std::transform(str.begin(), str.end(), str.begin(),
+                 [](char c) { return std::toupper(c, std::locale::classic()); });
+  return str;
+}
+
+DirectoryBlobReader::DiscContent::DiscContent(u64 offset, u64 size, ContentSource content_source)
+    : m_offset(offset), m_size(size), m_content_source(content_source)
+{
+}
+
+DirectoryBlobReader::DiscContent::DiscContent(u64 offset) : m_offset(offset), m_size(0)
```

Comment:
`m_size` can be initialized to zero in the class body. The other constructor will still work as intended.